### PR TITLE
Add new constructor to function parser

### PR DESCRIPTION
**Pull request by luca-heltai**

### CODE REVIEW

File: doc/news/changes/minor/20180721LucaHeltai [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Added new constructor type for FunctionParser class.
```
[the remark this answers is not in the document]
Comment (bangerth) [reply doc/news/changes/minor/20180721LucaHeltai]: We also use present tense most often: "Add new constructor..."

File: tests/base/function_parser_10.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// test rand function and expression only constructor
```
Comment (tjhei) [inline tests/base/function_parser_10.cc]: update comment please

**Comment by tjhei** [comment]
would you mind adding a check for an expression with more than one component? I am not 100% convinced that the ``string_split(expression).size()`` is working correctly (at least I had to spend time to convince myself).

File: source/base/function_parser.cc [hunk 3]
```diff
@@ -53,6 +54,33 @@ FunctionParser<dim>::FunctionParser(const unsigned int n_components,
...
+{
+  auto constants_map = Patterns::Tools::Convert<ConstMap>::to_value(
+    constants,
+    Patterns::Map(Patterns::Anything(),
+                  Patterns::Double(),
+                  0,
+                  Patterns::Map::max_int_value,
+                  ",",
+                  "=")
+      .clone());
```
Comment (tjhei) [inline source/base/function_parser.cc]: why do you need clone here?

**Comment by bangerth** [comment]
Oh, I goofed and merged even though @tjhei 's comments were still not addressed.

@luca-heltai -- can you please still address these?

File: source/base/function_parser.cc (hunk 3 above)
Comment (luca-heltai) [reply source/base/function_parser.cc]: because the function expects a std::unique_ptr.

Comment (bangerth) [reply source/base/function_parser.cc]: I think what @tjhei was thinking of was to use `std_cxx14::make_unique<Patterns::Map>(...arguments as above...)`.

**Pull request merged**